jobs/audit_jobs.py

```diff
@@ -0,0 +1,95 @@
+from jobs import jobs_utils
+from jobs.decorators import audit_decorators
+from jobs.transforms import base_model_audits
+from jobs.transforms import user_audits  # pylint: disable=unused-import
+import python_utils
+
+import apache_beam as beam
+
+AUDIT_DO_FNS_BY_KIND = audit_decorators.AuditsExisting.get_audits_by_kind()
+
+KIND_BY_INDEX = list(AUDIT_DO_FNS_BY_KIND.keys())
+AUDIT_DO_FNS_BY_INDEX = [AUDIT_DO_FNS_BY_KIND[k] for k in KIND_BY_INDEX]
+
+
+class AuditAllStorageModelsJob(base_jobs.JobBase):
+    """Runs a comprehensive audit on every model in storage."""
+
+    def run(self):
+        """Returns a PCollection of audit errors aggregated from all models."""
+        if self.job_options.model_getter is None:
+            raise ValueError('JobOptions.model_getter must not be None')
+
+        existing_models, deleted_models = (
+            self.pipeline
+            | 'Get all models' >> self.job_options.model_getter()
+            | 'Partition by model.deleted' >> (
+                beam.Partition(lambda model, _: int(model.deleted), 2))
+        )
+
+        models_of_kind_by_index = (
+            existing_models
+            # NOTE: Partition returns a statically-sized list of PCollections.
+            # Creating partitions can be wasteful when there are only a few
+            # items, like in our unit tests. In exchange, we can take advantage
+            # of the high parallelizability of PCollections, which are designed
+            # for huge datasets and parallel processing.
+            #
+            # Alternatively, we could have used GroupBy. However, that returns
+            # an _iterable_ over the groups instead of a PCollection, and so it
+            # is vulnerable to out-of-memory errors.
+            #
+            # Since this job is concerned with running audits on EVERY MODEL IN
+            # STORAGE, Partition is the clear winner regardless of the overhead
+            # we'll see in unit tests.
+            | 'Split models into parallelizable PCollections' >> beam.Partition(
+                lambda m, _, kinds: kinds.index(jobs_utils.get_model_kind(m)),
+                # NOTE: Partition requires a hard-coded number of slices, it
+                # cannot be used with dynamic numbers generated in a pipeline.
+                # The KIND_BY_INDEX list satisfies this requirement, since it
+                # is determined and finalized after importing all of the modules
+                # that use the AuditsExisting decorator.
+                len(KIND_BY_INDEX), KIND_BY_INDEX)
+        )
+
+        get_label = lambda audit, kind: 'Run %s on %s' % (audit.__name__, kind)
+        audit_error_pcolls = [
+            models_of_kind | get_label(audit, kind) >> beam.ParDo(audit())
+
+            for kind, audits, models_of_kind in python_utils.ZIP(
+                KIND_BY_INDEX, AUDIT_DO_FNS_BY_INDEX, models_of_kind_by_index)
+            if models_of_kind and audits
+            for audit in audits
+        ]
```

Comment:
Hmm, could we split this up into two separate for-cycles? I think that two fors in one list comprehension are a bit confusing.

Reply:
Yes, thanks.